### PR TITLE
Closing input stream, runner session in TensorFlowInferenceInterface.java and fixing bit changes

### PR DESCRIPTION
**Pull request by rajendraarora16**

### CODE REVIEW

**Comment by caisq** [comment]
@tensorflow-jenkins test this please

File: tensorflow/contrib/all_reduce/python/all_reduce.py [hunk 1]
```diff
@@ -317,7 +317,7 @@ def _build_ring_gather(input_tensors, devices, num_subchunks,
   num_devices = len(input_tensors)
   if num_devices == 0:
     return []
-  if num_devices == 1:
+  elif num_devices == 1:
```
Comment (asimshankar) [inline tensorflow/contrib/all_reduce/python/all_reduce.py]: Why change this? The existing version seems equivalent, no?

File: tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java [hunk 2]
```diff
@@ -264,6 +266,7 @@ public void close() {
     closeFeeds();
     closeFetches();
     sess.close();
+    runner.close();
```
Comment (asimshankar) [inline tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java]: Does this work? I don't think `Session.Runner` has a `close` method - see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/java/src/main/java/org/tensorflow/Session.java

File: tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java [hunk 3]
```diff
@@ -156,6 +156,8 @@ public TensorFlowInferenceInterface(InputStream is) {
       if (VERSION.SDK_INT >= 18) {
         Trace.endSection(); // initializeTensorFlow.
       }
+
+      is.close();
```
Comment (asimshankar) [inline tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java]: It might be better to invoke `is.close()` in the `finally` clause instead of here. That way any other exceptions raised in this block still result in the `InputStream` being closed.

File: tensorflow/contrib/crf/python/kernel_tests/crf_test.py [hunk 4]
```diff
@@ -114,7 +114,7 @@ def testCrfLogNorm(self):
         np.array([[4, 5, -3], [3, -1, 3], [-1, 2, 1], [0, 0, 0]],
                  dtype=np.float32),
         np.array([[3, -1, 3]],
-                 dtype=np.float32),
+                 dtype=np.float32)
```
Comment (asimshankar) [inline tensorflow/contrib/crf/python/kernel_tests/crf_test.py]: This change doesn't seem necessary?
In general, I'd suggest that we avoid these one-character formatting changes unless there are other more significant changes to the file (which isn't the case here).

**Review by asimshankar — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java (hunk 3 above)
Comment (rajendraarora16) [reply tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java]: Done!

File: tensorflow/contrib/all_reduce/python/all_reduce.py (hunk 1 above)
Comment (rajendraarora16) [reply tensorflow/contrib/all_reduce/python/all_reduce.py]: Two times `if-statement` was being shown using the same variable. That's why thought to put it in with `elif` statement. Should I revert it back?

Comment (asimshankar) [reply tensorflow/contrib/all_reduce/python/all_reduce.py]: Yes please. Not that there is anything wrong with the `elif`, but then again, not there was nothing wrong with the `if` either :) So in the interest of avoiding superfluous changes, let's revert this.

**Comment by rajendraarora16** [comment]
Changes done! Can we go ahead? :)

File: tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java [hunk 5]
```diff
@@ -158,6 +158,8 @@ public TensorFlowInferenceInterface(InputStream is) {
       }
     } catch (IOException e) {
       throw new RuntimeException("Failed to load model from the input stream", e);
+    } finally {
+       is.close();
```
Comment (asimshankar) [inline tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java]: Oh, wait, my apologies - I had misread this change.
It seems you're changing the contract of the public `TensorFlowInferenceInterface(InputStream)` constructor so that it "takes ownership" of the `InputStream`.

Could you elaborate on why we'd want to change the contract? It seems more reasonable to me that the creator of the `InputStream` be responsible for closing it. Am I missing something?

Comment (rajendraarora16) [reply tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java]: `It seems you're changing the contract of the public ` I am sorry I didn't get you. Regarding reasonable, `InputStream ` owns resources that must be explicitly freed by invoking `is.close()` and yes the creator of the `InputStream` be responsible for closing it.

Comment (asimshankar) [reply tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java]: Let me rephrase.

The constructor is `public TensorFlowInferenceInterface(InputStream is)`.
Without this PR, the constructor would not call `is.close()`. Which means that client code would look like this:

```java
InputStream s = createInputStream();
TensorFlowInferenceInterface tf = new TensorFlowInferenceInterface(s);
// Perhaps do something else with s
s.close()
```

In this PR, you're changing the constructor to call `is.close()`, which means that the creation and closing of the `InputStream` is asymmetric. For example:

```java
InputStream s = createInputStream();
TensorFlowInferenceInterface tf = new TensorFlowInferenceInterface(s);
// s can no longer be used by the client as it is  closed.
```

I don't think we should be making that change.

Comment (rajendraarora16) [reply tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java]: Ok got it!, just thought in case. In future, if we you are using that method. So it could be beneficial if we are closing it with final clause within the method itself. However, I don't think that is going to throw an error.

Comment (asimshankar) [reply tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java]: I didn't quite follow what you meant by "it could be beneficial if we are closing it with final clause within the method itself", but I gather we're on the same page, so closing this PR.

Thanks!
